src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionDescriptorCollectionProvider.cs

```diff
@@ -35,36 +49,48 @@ public ActionDescriptorCollection ActionDescriptors
         {
             get
             {
-                if (_collection == null)
+                var result = _collection;
+
+                if (result == null)
                 {
-                    _collection = GetCollection();
+                    result = GetCollection();
+                    _collection = result;
                 }
 
-                return _collection;
+                return result;
             }
         }
 
         private ActionDescriptorCollection GetCollection()
         {
-            var providers =
-                _serviceProvider.GetServices<IActionDescriptorProvider>()
-                                .OrderBy(p => p.Order)
-                                .ToArray();
-
             var context = new ActionDescriptorProviderContext();
 
-            foreach (var provider in providers)
+            for (var i = 0; i < _actionDescriptorProviders.Length; i++)
             {
-                provider.OnProvidersExecuting(context);
+                _actionDescriptorProviders[i].OnProvidersExecuting(context);
             }
 
-            for (var i = providers.Length - 1; i >= 0; i--)
+            for (var i = _actionDescriptorProviders.Length - 1; i >= 0; i--)
             {
-                providers[i].OnProvidersExecuted(context);
+                _actionDescriptorProviders[i].OnProvidersExecuted(context);
             }
 
             return new ActionDescriptorCollection(
-                new ReadOnlyCollection<ActionDescriptor>(context.Results), 0);
+                new ReadOnlyCollection<ActionDescriptor>(context.Results),
+                Interlocked.Increment(ref _version));
+        }
+
+        private void OnChange()
+        {
+            _collection = null;
```

Comment:
This isn't right. You need to get a new change token each time a change occurs. Please take a look at how this is done in configuration.

What I'd suggest:
1. Change comes in
2. Get new change tokens (don't subscribe yet)
3. Populate the collection eagerly
4. Subscribe to change tokens

Doing the handshake like this ensures that you don't have any race conditions. We might want to use a composite change token here, I believe that's what configuration does.